### PR TITLE
Add Container widget

### PR DESCRIPTION
**Pull request by zaynetro**
Creating this WIP PR early as I have encountered questions I would like some feedback on.

Background and padding seems straightforward at the moment. 

My main struggle was border. When stroking the shape we can draw outside of the widget constraints. To compensate that I needed to shrink the child widget box constraints.

I've noticed that `Button` has the same issue with 1px being outside of given box constraints.

Current implementation:

image

No extra logic (this is what happens with `Button`):

image

My question is what is the right way to tackle this issue?

### CODE REVIEW

**Comment by raphlinus** [comment]
The "foreground" of a widget should be inside the layout bounds, though this is a bit of a subjective criterion. I do want to have more sophisticated tracking of paint bounds, but right now it's loose; you can assume you are *allowed* to paint anywhere. We're probably not systematic about this in our existing bordered widgets (button, textbox, etc), but it's probably not a huge deal because those borders tend to be thin.

**Comment by zaynetro** [comment]
Thanks for the comments! I plan to get back to this MR by the end of October. Currently overwhelmed at work...

**Comment by zaynetro** [comment]
OK. I think current container is already good enough. Let me know if you think something else must be included.

**Comment by zaynetro** [comment]
Good to go?

File: src/widget/container.rs [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+    pub fn empty(self) -> impl Widget<T> {
+        self.child(SizedBox::empty())
+    }
+
+    pub fn color(mut self, brush: impl Into<PaintBrush>) -> Self {
+        self.style.color = Some(brush.into());
+        self
+    }
+
+    pub fn border(mut self, brush: impl Into<PaintBrush>, width: f64) -> Self {
```
Comment (cmyr) [inline src/widget/container.rs]: I think we need to either let the user set the stroke style or else not include it in the type.

File: src/widget/container.rs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+use crate::shell::kurbo::{Point, Rect, Size};
+use crate::shell::piet::{PaintBrush, RenderContext, StrokeStyle};
+use crate::widget::Padding;
+use crate::widget::SizedBox;
+use crate::{
+    BaseState, BoxConstraints, Data, Env, Event, EventCtx, LayoutCtx, PaintCtx, UpdateCtx, Widget,
+};
+
+use std::marker::PhantomData;
```
[the remark this answers is not in the document]
Comment (zaynetro) [reply src/widget/container.rs]: I was under impression that rustfmt does that by default with `reorder_imports`. Let's see.

File: src/widget/container.rs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+            self.border_padding(),
+            ContainerRaw::new(self.style, Padding::uniform(self.padding, inner)),
+        )
+    }
+
+    pub fn empty(self) -> impl Widget<T> {
+        self.child(SizedBox::empty())
+    }
+
+    pub fn color(mut self, brush: impl Into<PaintBrush>) -> Self {
```
[the remark this answers is not in the document]
Comment (zaynetro) [reply src/widget/container.rs]: Called it `color` as in Flutter but I agree that `background` is more descriptive.

File: src/widget/container.rs (hunk 1 above)
Comment (zaynetro) [reply src/widget/container.rs]: I will drop the stroke for now. Once we have a need it could be added.

**Comment by zaynetro** [comment]
`ContainerRaw` is basically simulating what `TextBoxRaw` or `CheckboxRaw` is doing. I followed the mechanism that we already have. Although, I agree that the constructing might need some improvements and I will think about those.

Thanks for the comments!

File: src/widget/container.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (zaynetro) [reply src/widget/container.rs]: > Reorder import and extern crate statements alphabetically in groups (a group is separated by a newline).

So if imports are separated by newlines then they will not be orderred. This is a bit annoying.

**Comment by cmyr** [comment]
I hadn't really looked at checkbox, so thanks for that; I probably would have had similar questions if I'd been doing the review for that PR. In the case of checkbox it seems like this pattern is particularly curious, since it would be just as easy to impl `Widget` on the `Checkbox` struct itself.

In any case, I would not consider this to be a general pattern to be followed in most cases; it should be a pattern we pick for specific reasons related to the constraints of the implementation.

So I think this is okay if you think it's the best approach, but you should not feel like this pattern is a requirement.

I _would_ like some documentation for all public items. If you're not sure about what these docs should be feel free to ping me on zulip and we can talk about it.

Let me know when you think this is ready for another look!

**Comment by Felk** [comment]
Regarding the inside/outside drawing issue: Having dealt with various CSS styling and layouting headaches on the web, I've personally come to the following conclusions:

- Side effects to layouting shall always be contained within the widget. E.g. if A and B are 100px wide each, but B has a padding of 20px, B should _not_ be 140px wide total, but stay at 100px. Anything else is imo unintuitive and makes layouting harder. Have a look at the documentation of the CSS-property [box-sizing](https://developer.mozilla.org/en-US/docs/Web/CSS/box-sizing) for more information.
- Now this isn't directly relevant to the border-drawing issue, because just _drawing_ outside of the bounds wouldn't affect layouting. Nontheless I'd expect a border to behave similarly to a padding and stay within the bounds.
- Also, for people being familiar with CSS, the behaviour of drawing _outside_ of the element's bounds will probably be intuitively known as an [outline](https://developer.mozilla.org/en-US/docs/Web/CSS/outline) instead of a border.

**Comment by zaynetro** [comment]
OK. I removed `ContainerRaw` in favour of just one struct. There is one question that I am not sure about. Otherwise, I think it is in its final shape.

@cmyr

File: druid/src/widget/container.rs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    fn paint(&mut self, paint_ctx: &mut PaintCtx, base_state: &BaseState, data: &T, env: &Env) {
+        // Add border offset
+        if let Some(ref border) = self.style.border {
+            let offset = border.width / 2.0;
+            paint_ctx.transform(Affine::translate((offset, offset)));
+        }
+
+        // Paint background color
+        if let Some(ref brush) = self.style.background {
+            let rect = Rect::from_origin_size(Point::ZERO, base_state.size());
```
[the remark this answers is not in the document]
Comment (zaynetro) [reply druid/src/widget/container.rs]: Changed to use WidgetPod. Most of the other widgets seem to use it.

File: druid/src/widget/container.rs [hunk 5]
```diff
@@ -114,25 +73,46 @@ impl<T: Data> Widget<T> for ContainerRaw<T> {
...
-        self.inner.paint(paint_ctx, base_state, data, env);
+        self.inner.paint_with_offset(paint_ctx, data, env);
     }
 
     fn layout(&mut self, ctx: &mut LayoutCtx, bc: &BoxConstraints, data: &T, env: &Env) -> Size {
-        self.inner.layout(ctx, bc, data, env)
+        bc.debug_check("Container");
+
+        // Shrink constraints by border offset
+        let border_width = match self.style.border {
```
Comment (cmyr) [inline druid/src/widget/container.rs]: alternatively,

```rust
self.style.border.as_ref().map(|b| b.width).unwrap_or_default();
```

(not necessary to change, just wanted to point out the one-liner)

**Pull request merged**